File: backend/src/services/events.rs
use chrono::Utc;
use sea_orm::*;
use std::error::Error;

use crate::entity::prelude::*;
use crate::entity::{
    attendance, attendee, event, host, user, AttendanceStatus, EventStatus, EventType,
};
use crate::schemas::event::{
    AttendanceProfile, EventAttendeeDetails, EventAttendanceSummary, PhysicalEventResponse,
};

/// Helper to calculate distance between two coordinates using Haversine formula
/// Returns distance in meters
pub fn calculate_distance(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    const EARTH_RADIUS_KM: f64 = 6371.0;

    let lat1_rad = lat1.to_radians();
    let lat2_rad = lat2.to_radians();
    let delta_lat = (lat2 - lat1).to_radians();
    let delta_lon = (lon2 - lon1).to_radians();

    let a = (delta_lat / 2.0).sin().powi(2)
        + lat1_rad.cos() * lat2_rad.cos() * (delta_lon / 2.0).sin().powi(2);
    let c = 2.0 * a.sqrt().atan2((1.0 - a).sqrt());

    EARTH_RADIUS_KM * c * 1000.0 // Convert to meters
}

/// Create a new physical event
pub async fn create_physical_event(
    db: &DatabaseConnection,
    host_id: i32,
    name: String,
    description: Option<String>,
    category: crate::entity::EventCategory,
    visibility: crate::entity::EventVisibility,
    start_time: chrono::DateTime<Utc>,
    end_time: chrono::DateTime<Utc>,
    latitude: f64,
    longitude: f64,
    geofence_radius: Option<f64>,
    attendance_profile: AttendanceProfile,
) -> Result<event::Model, Box<dyn Error>> {
    // Create location string from coordinates
    let location = format!("{},{}", latitude, longitude);

    // Store geofence and attendance settings in description metadata
    // In production, you'd want separate fields in the database
    let metadata = serde_json::json!({
        "latitude": latitude,
        "longitude": longitude,
        "geofence_radius": geofence_radius,
        "attendance_profile": attendance_profile,
        "attendance_window_minutes": attendance_profile.duration_minutes(),
    });

    let description_with_metadata = description
        .map(|d| format!("{}\n__METADATA__:{}", d, metadata.to_string()))
        .unwrap_or_else(|| format!("__METADATA__:{}", metadata.to_string()));

    let new_event = event::ActiveModel {
        title: Set(name),
        description: Set(description_with_metadata),
        location: Set(location),
        event_type: Set(EventType::Physical),
        category: Set(category),
        status: Set(EventStatus::Scheduled),
        visibility: Set(visibility),
        host_id: Set(host_id),
        start_time: Set(start_time),
        end_time: Set(end_time),
        created_at: Set(Utc::now()),
        updated_at: Set(Utc::now()),
        ..Default::default()
    };

    let event = new_event.insert(db).await?;
    Ok(event)
}

/// Get event by ID with host information
pub async fn get_event_by_id(
    db: &DatabaseConnection,
    event_id: i32,
) -> Result<(event::Model, user::Model), Box<dyn Error>> {
    let event = Event::find_by_id(event_id)
        .one(db)
        .await?
        .ok_or("Event not found")?;

    // Get host information
    let host_record = Host::find()
        .filter(host::Column::UserId.eq(event.host_id))
        .one(db)
        .await?
        .ok_or("Host not found")?;

    let host_user = User::find_by_id(host_record.user_id)
        .one(db)
        .await?
        .ok_or("Host user not found")?;

    Ok((event, host_user))
}

/// Update an existing event
pub async fn update_event(
    db: &DatabaseConnection,
    event_id: i32,
    host_id: i32,
    name: Option<String>,
    description: Option<String>,
    category: Option<crate::entity::EventCategory>,
    visibility: Option<crate::entity::EventVisibility>,
    start_time: Option<chrono::DateTime<Utc>>,
    end_time: Option<chrono::DateTime<Utc>>,
    latitude: Option<f64>,
    longitude: Option<f64>,
    geofence_radius: Option<f64>,
    attendance_profile: Option<AttendanceProfile>,
) -> Result<event::Model, Box<dyn Error>> {
    // Verify event exists and belongs to host
    let existing_event = Event::find_by_id(event_id)
        .one(db)
        .await?
        .ok_or("Event not found")?;

    if existing_event.host_id != host_id {
        return Err("Unauthorized: You are not the host of this event".into());
    }

    // Parse existing metadata
    let mut metadata: serde_json::Value = if let Some(meta_str) = existing_event
        .description
        .split("__METADATA__:")
        .nth(1)
    {
        serde_json::from_str(meta_str).unwrap_or(serde_json::json!({}))
    } else {
        serde_json::json!({})
    };

    // Update metadata fields
    if let Some(lat) = latitude {
        metadata["latitude"] = serde_json::json!(lat);
    }
    if let Some(lon) = longitude {
        metadata["longitude"] = serde_json::json!(lon);
    }
    if let Some(radius) = geofence_radius {
        metadata["geofence_radius"] = serde_json::json!(radius);
    }
    if let Some(profile) = attendance_profile {
        metadata["attendance_profile"] = serde_json::json!(profile);
        metadata["attendance_window_minutes"] = serde_json::json!(profile.duration_minutes());
    }

    // Update location if coordinates changed
    let new_location = if latitude.is_some() || longitude.is_some() {
        let lat = latitude.unwrap_or(metadata["latitude"].as_f64().unwrap_or(0.0));
        let lon = longitude.unwrap_or(metadata["longitude"].as_f64().unwrap_or(0.0));
        Some(format!("{},{}", lat, lon))
    } else {
        None
    };

    // Update description with metadata
    let new_description = if description.is_some() || latitude.is_some() || longitude.is_some() || geofence_radius.is_some() || attendance_profile.is_some() {
        let desc = description.unwrap_or_else(|| {
            existing_event
                .description
                .split("__METADATA__:")
                .next()
                .unwrap_or("")
                .to_string()
        });
        Some(format!("{}\n__METADATA__:{}", desc, metadata.to_string()))
    } else {
        None
    };

    let mut event: event::ActiveModel = existing_event.into();

    if let Some(n) = name {
        event.title = Set(n);
    }
    if let Some(d) = new_description {
        event.description = Set(d);
    }
    if let Some(c) = category {
        event.category = Set(c);
    }
    if let Some(v) = visibility {
        event.visibility = Set(v);
    }
    if let Some(st) = start_time {
        event.start_time = Set(st);
    }
    if let Some(et) = end_time {
        event.end_time = Set(et);
    }
    if let Some(loc) = new_location {
        event.location = Set(loc);
    }
    event.updated_at = Set(Utc::now());

    let updated_event = event.update(db).await?;
    Ok(updated_event)
}

/// Cancel an event
pub async fn cancel_event(
    db: &DatabaseConnection,
    event_id: i32,
    host_id: i32,
) -> Result<event::Model, Box<dyn Error>> {
    let existing_event = Event::find_by_id(event_id)
        .one(db)
        .await?
        .ok_or("Event not found")?;

    if existing_event.host_id != host_id {
        return Err("Unauthorized: You are not the host of this event".into());
    }

    if existing_event.status == EventStatus::Cancelled {
        return Err("Event is already cancelled".into());
    }

    let mut event: event::ActiveModel = existing_event.into();
    event.status = Set(EventStatus::Cancelled);
    event.updated_at = Set(Utc::now());

    let cancelled_event = event.update(db).await?;
    Ok(cancelled_event)
}

/// RSVP for an event
pub async fn rsvp_event(
    db: &DatabaseConnection,
    event_id: i32,
    attendee_id: i32,
) -> Result<attendance::Model, Box<dyn Error>> {
    // Verify event exists
    let event = Event::find_by_id(event_id)
        .one(db)
        .await?
        .ok_or("Event not found")?;

    if event.status == EventStatus::Cancelled {
        return Err("Cannot RSVP to a cancelled event".into());
    }

    if event.status == EventStatus::Completed {
        return Err("Cannot RSVP to a completed event".into());
    }

    // Check if already registered
    if let Some(_existing) = Attendance::find()
        .filter(attendance::Column::EventId.eq(event_id))
        .filter(attendance::Column::AttendeeId.eq(attendee_id))
        .one(db)
        .await?
    {
        return Err("Already registered for this event".into());
    }

    // Create attendance record with Registered status
    let new_attendance = attendance::ActiveModel {
        event_id: Set(event_id),
        attendee_id: Set(attendee_id),
        status: Set(AttendanceStatus::Registered),
        created_at: Set(Utc::now()),
        updated_at: Set(Utc::now()),
        ..Default::default()
    };

    let attendance_record = new_attendance.insert(db).await?;
    Ok(attendance_record)
}

/// Mark attendance (check-in)
pub async fn mark_attendance(
    db: &DatabaseConnection,
    event_id: i32,
    attendee_id: i32,
    verify_location: bool,
    user_latitude: Option<f64>,
    user_longitude: Option<f64>,
) -> Result<(attendance::Model, bool, bool), Box<dyn Error>> {
    // Get event
    let event = Event::find_by_id(event_id)
        .one(db)
        .await?
        .ok_or("Event not found")?;

    // Parse event metadata
    let metadata: serde_json::Value = if let Some(meta_str) =
        event.description.split("__METADATA__:").nth(1)
    {
        serde_json::from_str(meta_str).unwrap_or(serde_json::json!({}))
    } else {
        return Err("Event metadata not found".into());
    };

    // Get attendance record
    let attendance_record = Attendance::find()
        .filter(attendance::Column::EventId.eq(event_id))
        .filter(attendance::Column::AttendeeId.eq(attendee_id))
        .one(db)
        .await?
        .ok_or("No RSVP found. Please RSVP first.")?;

    if attendance_record.status == AttendanceStatus::CheckedIn {
        return Err("Already checked in for this event".into());
    }

    // Verify timing
    let now = Utc::now();
    if now < event.start_time {
        return Err("Event has not started yet".into());
    }

    // Check attendance window
    let attendance_window_minutes = metadata["attendance_window_minutes"].as_i64();
    let mut is_late = false;

    if let Some(window_minutes) = attendance_window_minutes {
        let window_end = event.start_time + chrono::Duration::minutes(window_minutes);
        if now > window_end {
            if event.status != EventStatus::Completed {
                // After window but event not closed - flag as late
                is_late = true;
            } else {
                return Err("Attendance window has closed".into());
            }
        }
    } else {
        // Unlimited - check if event ended
        if now > event.end_time && event.status == EventStatus::Completed {
            return Err("Event has ended and been closed".into());
        } else if now > event.end_time {
            is_late = true;
        }
    }

    // Verify location if required
    let mut location_verified = false;
    if verify_location {
        let user_lat = user_latitude.ok_or("Latitude required for location verification")?;
        let user_lon = user_longitude.ok_or("Longitude required for location verification")?;

        let event_lat = metadata["latitude"]
            .as_f64()
            .ok_or("Event latitude not found")?;
        let event_lon = metadata["longitude"]
            .as_f64()
            .ok_or("Event longitude not found")?;
        let geofence_radius = metadata["geofence_radius"].as_f64().unwrap_or(100.0);

        let distance = calculate_distance(user_lat, user_lon, event_lat, event_lon);

        if distance > geofence_radius {
            return Err(format!(
                "You are too far from the event location ({:.0}m away, max: {:.0}m)",
                distance, geofence_radius
            )
            .into());
        }

        location_verified = true;
    }

    // Update attendance status
    let mut attendance: attendance::ActiveModel = attendance_record.into();
    attendance.status = Set(AttendanceStatus::CheckedIn);
    attendance.updated_at = Set(now);

    let updated_attendance = attendance.update(db).await?;
    Ok((updated_attendance, location_verified, is_late))
}

/// Get event attendance summary for host
pub async fn get_event_attendance_summary(
    db: &DatabaseConnection,
    event_id: i32,
    host_id: i32,
) -> Result<EventAttendanceSummary, Box<dyn Error>> {
    // Verify event belongs to host
    let event = Event::find_by_id(event_id)
        .one(db)
        .await?
        .ok_or("Event not found")?;

    if event.host_id != host_id {
        return Err("Unauthorized: You are not the host of this event".into());
    }

    // Get metadata for attendance window
    let metadata: serde_json::Value = if let Some(meta_str) =
        event.description.split("__METADATA__:").nth(1)
    {
        serde_json::from_str(meta_str).unwrap_or(serde_json::json!({}))
    } else {
        serde_json::json!({})
    };

    let attendance_window_minutes = metadata["attendance_window_minutes"].as_i64();

    // Get all attendance records with attendee and user info
    let attendance_records = Attendance::find()
        .filter(attendance::Column::EventId.eq(event_id))
        .find_also_related(Attendee)
        .all(db)
        .await?;

    let mut attendees = Vec::new();
    let mut total_rsvp = 0;
    let mut total_checked_in = 0;
    let mut total_no_show = 0;
    let mut total_late = 0;

    for (attendance_rec, attendee_opt) in attendance_records {
        total_rsvp += 1;

        match attendance_rec.status {
            AttendanceStatus::CheckedIn => total_checked_in += 1,
            AttendanceStatus::NoShow => total_no_show += 1,
            _ => {}
        }

        if let Some(attendee_rec) = attendee_opt {
            // Get user info
            let user_info = User::find_by_id(attendee_rec.user_id).one(db).await?;

            if let Some(user) = user_info {
                // Check if late
                let mut is_late = false;
                if attendance_rec.status == AttendanceStatus::CheckedIn {
                    if let Some(window_minutes) = attendance_window_minutes {
                        let window_end =
                            event.start_time + chrono::Duration::minutes(window_minutes);
                        if attendance_rec.updated_at > window_end {
                            is_late = true;
                            total_late += 1;
                        }
                    }
                }

                attendees.push(EventAttendeeDetails {
                    attendee_id: attendee_rec.id,
                    username: user.username,
                    first_name: user.first_name,
                    last_name: user.last_name,
                    status: attendance_rec.status,
                    rsvp_time: attendance_rec.created_at,
                    check_in_time: if attendance_rec.status == AttendanceStatus::CheckedIn {
                        Some(attendance_rec.updated_at)
                    } else {
                        None
                    },
                    is_late,
                });
            }
        }
    }

    Ok(EventAttendanceSummary {
        event_id,
        total_rsvp,
        total_checked_in,
        total_no_show,
        total_late,
        attendees,
    })
}

/// List physical events (with optional filters)
pub async fn list_physical_events(
    db: &DatabaseConnection,
    status: Option<EventStatus>,
    host_id: Option<i32>,
    limit: Option<u64>,
    offset: Option<u64>,
) -> Result<(Vec<event::Model>, u64), Box<dyn Error>> {
    let mut query = Event::find().filter(event::Column::EventType.eq(EventType::Physical));

    if let Some(s) = status {
        query = query.filter(event::Column::Status.eq(s));
    }

    if let Some(h) = host_id {
        query = query.filter(event::Column::HostId.eq(h));
    }

    // Get total count
    let total = query.clone().count(db).await?;

    // Apply pagination
    if let Some(l) = limit {
        query = query.limit(l);
    }
    if let Some(o) = offset {
        query = query.offset(o);
    }

    let events = query.all(db).await?;
    Ok((events, total))
}

/// Convert event model to response with metadata
pub fn event_to_response(
    event: event::Model,
    host_name: String,
    rsvp_count: i32,
    checked_in_count: i32,
) -> Result<PhysicalEventResponse, Box<dyn Error>> {
    // Parse metadata
    let metadata: serde_json::Value = if let Some(meta_str) =
        event.description.split("__METADATA__:").nth(1)
    {
        serde_json::from_str(meta_str).unwrap_or(serde_json::json!({}))
    } else {
        return Err("Event metadata not found".into());
    };

    let description = event
        .description
        .split("__METADATA__:")
        .next()
        .unwrap_or("")
        .to_string();

    let description_cleaned = if description.is_empty() {
        None
    } else {
        Some(description)
    };

    // Extract attendance profile
    let attendance_profile: AttendanceProfile = serde_json::from_value(
        metadata["attendance_profile"].clone(),
    )
    .unwrap_or(AttendanceProfile::Standard);

    Ok(PhysicalEventResponse {
        id: event.id,
        name: event.title,
        description: description_cleaned,
        category: event.category,
        status: event.status,
        visibility: event.visibility,
        host_id: event.host_id,
        host_name,
        start_time: event.start_time,
        end_time: event.end_time,
        latitude: metadata["latitude"].as_f64().unwrap_or(0.0),
        longitude: metadata["longitude"].as_f64().unwrap_or(0.0),
        geofence_radius: metadata["geofence_radius"].as_f64(),
        attendance_profile,
        attendance_window_minutes: metadata["attendance_window_minutes"].as_i64().map(|v| v as i32),
        rsvp_count,
        checked_in_count,
        created_at: event.created_at,
        updated_at: event.updated_at,
    })
}

File: backend/app/services/events.py
import datetime
import logging
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from app.models import (
    Attendance,
    AttendanceStatus,
    Attendee,
    Event,
    EventCategory,
    EventStatus,
    EventType,
    EventVisibility,
    Host,
    BaseUser,
)
from app.schemas import (
    AttendanceProfile,
    EventAttendanceSummary,
    EventAttendeeDetails,
    PhysicalEventResponse,
    RecurrenceRule,
)
from app.utils.geo import calculate_distance

logger = logging.getLogger("app.services.events")


async def get_attendance_counts(
    db: AsyncSession, event_id: int
) -> tuple[int, int]:
    """Retrieve aggregate RSVP and Checked-In attendance counts for an event."""
    rsvp_stmt = select(func.count(Attendance.id)).filter(
        Attendance.event_id == event_id
    )
    checked_stmt = select(func.count(Attendance.id)).filter(
        Attendance.event_id == event_id,
        Attendance.status == AttendanceStatus.CheckedIn,
    )

    rsvp_res = await db.execute(rsvp_stmt)
    checked_res = await db.execute(checked_stmt)

    return rsvp_res.scalar() or 0, checked_res.scalar() or 0


async def create_physical_event(
    db: AsyncSession,
    host_id: int,
    name: str,
    description: str | None,
    category: EventCategory,
    visibility: EventVisibility,
    start_time: datetime.datetime,
    end_time: datetime.datetime,
    latitude: float,
    longitude: float,
    geofence_radius: float | None,
    attendance_profile: AttendanceProfile,
    recurrence: RecurrenceRule | None,
) -> Event:
    """Create a new physical event with geofencing coordinates and recurrence metadata."""
    # Location coordinates string representation
    location_str = f"{latitude},{longitude}"

    # Build legacy description metadata structure
    metadata = {
        "latitude": latitude,
        "longitude": longitude,
        "geofence_radius": geofence_radius if geofence_radius is not None else 100.0,
        "attendance_profile": attendance_profile.value,
        "attendance_window_minutes": attendance_profile.duration_minutes(),
        "recurrence": recurrence.model_dump() if recurrence else None,
    }

    # Format description to include metadata block
    clean_desc = description.strip() if description else ""
    event = Event(
        title=name,
        description=clean_desc,  # Will serialize metadata via helper property
        location=location_str,
        event_type="Physical",
        category=category,
        status=EventStatus.Scheduled,
        visibility=visibility,
        host_id=host_id,
        start_time=start_time,
        end_time=end_time,
    )

    # Force setter to write metadata block to description
    event.parsed_metadata = metadata

    db.add(event)
    await db.commit()
    await db.refresh(event)

    return event


async def get_event_by_id(db: AsyncSession, event_id: int) -> tuple[Event, BaseUser]:
    """Retrieve event by ID alongside its host BaseUser profile details."""
    stmt = select(Event).filter(Event.id == event_id)
    res = await db.execute(stmt)
    event = res.scalars().first()

    if not event:
        raise ValueError("Event not found")

    host_stmt = select(BaseUser).filter(BaseUser.id == event.host_id)
    host_res = await db.execute(host_stmt)
    host_user = host_res.scalars().first()

    if not host_user:
        raise ValueError("Host user not found")

    return event, host_user


async def update_event(
    db: AsyncSession,
    event_id: int,
    host_id: int,
    name: str | None,
    description: str | None,
    category: EventCategory | None,
    visibility: EventVisibility | None,
    start_time: datetime.datetime | None,
    end_time: datetime.datetime | None,
    latitude: float | None,
    longitude: float | None,
    geofence_radius: float | None,
    attendance_profile: AttendanceProfile | None,
    recurrence: RecurrenceRule | None,
) -> Event:
    """Update event properties and internal coordinate metadata blocks."""
    event, _ = await get_event_by_id(db, event_id)

    if event.host_id != host_id:
        raise PermissionError("Unauthorized: You are not the host of this event")

    # Load current metadata properties
    metadata = event.parsed_metadata

    if latitude is not None:
        metadata["latitude"] = latitude
    if longitude is not None:
        metadata["longitude"] = longitude
    if geofence_radius is not None:
        metadata["geofence_radius"] = geofence_radius
    if attendance_profile is not None:
        metadata["attendance_profile"] = attendance_profile.value
        metadata["attendance_window_minutes"] = (
            attendance_profile.duration_minutes()
        )
    if recurrence is not None:
        metadata["recurrence"] = recurrence.model_dump()

    # Update description text
    clean_desc = (
        description
        if description is not None
        else (event.clean_description or "")
    )
    event.description = clean_desc
    event.parsed_metadata = metadata

    if name is not None:
        event.title = name
    if category is not None:
        event.category = category
    if visibility is not None:
        event.visibility = visibility
    if start_time is not None:
        event.start_time = start_time
    if end_time is not None:
        event.end_time = end_time

    if latitude is not None or longitude is not None:
        lat = latitude if latitude is not None else metadata.get("latitude", 0.0)
        lon = (
            longitude if longitude is not None else metadata.get("longitude", 0.0)
        )
        event.location = f"{lat},{lon}"

    event.updated_at = datetime.datetime.utcnow()
    await db.commit()
    await db.refresh(event)

    return event


async def cancel_event(db: AsyncSession, event_id: int, host_id: int) -> Event:
    """Transition scheduled event status to Cancelled."""
    event, _ = await get_event_by_id(db, event_id)

    if event.host_id != host_id:
        raise PermissionError("Unauthorized: You are not the host of this event")

    if event.status == EventStatus.Cancelled:
        raise ValueError("Event is already cancelled")

    event.status = EventStatus.Cancelled
    event.updated_at = datetime.datetime.utcnow()

    await db.commit()
    await db.refresh(event)

    return event


async def rsvp_event(
    db: AsyncSession, event_id: int, attendee_id: int
) -> Attendance:
    """Create a new RSVPs (Registered) record for an upcoming event."""
    event, _ = await get_event_by_id(db, event_id)

    if event.status == EventStatus.Cancelled:
        raise ValueError("Cannot RSVP to a cancelled event")
    if event.status == EventStatus.Completed:
        raise ValueError("Cannot RSVP to a completed event")

    # Check for existing RSVP
    stmt = select(Attendance).filter(
        Attendance.event_id == event_id,
        Attendance.attendee_id == attendee_id,
    )
    res = await db.execute(stmt)
    existing = res.scalars().first()

    if existing:
        raise ValueError("Already registered for this event")

    rsvp = Attendance(
        event_id=event_id,
        attendee_id=attendee_id,
        status=AttendanceStatus.Registered,
    )

    db.add(rsvp)
    await db.commit()
    await db.refresh(rsvp)

    return rsvp


async def mark_attendance(
    db: AsyncSession,
    event_id: int,
    attendee_id: int,
    verify_location: bool,
    user_latitude: float | None,
    user_longitude: float | None,
) -> tuple[Attendance, bool, bool]:
    """Execute event attendee check-in and location/timing validations."""
    event, _ = await get_event_by_id(db, event_id)
    metadata = event.parsed_metadata

    # Retrieve RSVP
    stmt = select(Attendance).filter(
        Attendance.event_id == event_id,
        Attendance.attendee_id == attendee_id,
    )
    res = await db.execute(stmt)
    rsvp = res.scalars().first()

    if not rsvp:
        raise ValueError("No RSVP found. Please RSVP first.")

    if rsvp.status == AttendanceStatus.CheckedIn:
        raise ValueError("Already checked in for this event")

    now = datetime.datetime.utcnow()
    # Align timezone naive datetime for comparison
    start_time = event.start_time.replace(tzinfo=None)
    end_time = event.end_time.replace(tzinfo=None)

    if now < start_time:
        raise ValueError("Event has not started yet")

    # Timing rules & late flags
    window_minutes = metadata.get("attendance_window_minutes")
    is_late = False

    if window_minutes is not None:
        window_end = start_time + datetime.timedelta(minutes=window_minutes)
        if now > window_end:
            if event.status != EventStatus.Completed:
                is_late = True
            else:
                raise ValueError("Attendance window has closed")
    else:
        if now > end_time and event.status == EventStatus.Completed:
            raise ValueError("Event has ended and been closed")
        elif now > end_time:
            is_late = True

    # Location rules & geofencing
    location_verified = False
    if verify_location:
        if user_latitude is None or user_longitude is None:
            raise ValueError("Coordinates required for location verification")

        event_lat = metadata.get("latitude", 0.0)
        event_lon = metadata.get("longitude", 0.0)
        geofence_radius = metadata.get("geofence_radius", 100.0)

        distance = calculate_distance(
            user_latitude, user_longitude, event_lat, event_lon
        )

        if distance > geofence_radius:
            raise ValueError(
                f"You are too far from the event location ({distance:.0f}m away, max: {geofence_radius:.0f}m)"
            )

        location_verified = True

    rsvp.status = AttendanceStatus.CheckedIn
    rsvp.updated_at = now

    await db.commit()
    await db.refresh(rsvp)

    return rsvp, location_verified, is_late


async def get_event_attendance_summary(
    db: AsyncSession, event_id: int, host_id: int
) -> EventAttendanceSummary:
    """Retrieve attendance RSVP registries and check-in aggregates for the host."""
    event, _ = await get_event_by_id(db, event_id)

    if event.host_id != host_id:
        raise PermissionError("Unauthorized: You are not the host of this event")

    window_minutes = event.parsed_metadata.get("attendance_window_minutes")
    start_time = event.start_time.replace(tzinfo=None)

    # Get RSVP list
    stmt = (
        select(Attendance, BaseUser)
        .join(BaseUser, BaseUser.id == Attendance.attendee_id)
        .filter(Attendance.event_id == event_id)
    )
    res = await db.execute(stmt)
    rows = res.all()

    attendees = []
    total_rsvp = 0
    total_checked_in = 0
    total_no_show = 0
    total_late = 0

    for attendance_rec, user_rec in rows:
        total_rsvp += 1
        if attendance_rec.status == AttendanceStatus.CheckedIn:
            total_checked_in += 1
        elif attendance_rec.status == AttendanceStatus.NoShow:
            total_no_show += 1

        is_late = False
        if attendance_rec.status == AttendanceStatus.CheckedIn:
            check_in_time = attendance_rec.updated_at.replace(tzinfo=None)
            if window_minutes is not None:
                window_end = start_time + datetime.timedelta(
                    minutes=window_minutes
                )
                if check_in_time > window_end:
                    is_late = True
                    total_late += 1

        attendees.append(
            EventAttendeeDetails(
                attendee_id=user_rec.id,
                username=user_rec.username,
                first_name=user_rec.first_name,
                last_name=user_rec.last_name,
                status=attendance_rec.status,
                rsvp_time=attendance_rec.created_at,
                check_in_time=(
                    attendance_rec.updated_at
                    if attendance_rec.status == AttendanceStatus.CheckedIn
                    else None
                ),
                is_late=is_late,
            )
        )

    return EventAttendanceSummary(
        event_id=event_id,
        total_rsvp=total_rsvp,
        total_checked_in=total_checked_in,
        total_no_show=total_no_show,
        total_late=total_late,
        attendees=attendees,
    )


async def list_events_by_type(
    db: AsyncSession,
    event_type: EventType,
    status: EventStatus | None,
    host_id: int | None,
    limit: int | None,
    offset: int | None,
) -> tuple[list[Event], int]:
    """Retrieve events list by type with dynamic pagination and filter flags."""
    
    stmt = select(Event).filter(Event.event_type == event_type)  # <-- uses param now
    
    if status is not None:
        stmt = stmt.filter(Event.status == status)
    if host_id is not None:
        stmt = stmt.filter(Event.host_id == host_id)

    count_stmt = select(func.count()).select_from(stmt.subquery())
    count_res = await db.execute(count_stmt)
    total = count_res.scalar() or 0

    if offset is not None:
        stmt = stmt.offset(offset)
    if limit is not None:
        stmt = stmt.limit(limit)

    res = await db.execute(stmt)
    events = list(res.scalars().all())
    return events, total

async def event_to_response(
    db: AsyncSession, event: Event, host_name: str
) -> PhysicalEventResponse:
    """Compile event object model and attendance aggregates into PhysicalEventResponse."""
    rsvp_count, checked_in_count = await get_attendance_counts(db, event.id)
    metadata = event.parsed_metadata

    profile_val = metadata.get("attendance_profile", "standard")
    try:
        attendance_profile = AttendanceProfile(profile_val)
    except ValueError:
        attendance_profile = AttendanceProfile.Standard

    return PhysicalEventResponse(
        id=event.id,
        name=event.title,
        description=event.clean_description,
        category=event.category,
        status=event.status,
        visibility=event.visibility,
        host_id=event.host_id,
        host_name=host_name,
        start_time=event.start_time,
        end_time=event.end_time,
        latitude=metadata.get("latitude", 0.0),
        longitude=metadata.get("longitude", 0.0),
        geofence_radius=metadata.get("geofence_radius"),
        attendance_profile=attendance_profile,
        attendance_window_minutes=metadata.get("attendance_window_minutes"),
        rsvp_count=rsvp_count,
        checked_in_count=checked_in_count,
        created_at=event.created_at,
        updated_at=event.updated_at,
    )
